torch_geometric/transforms/chronological_split.py

```diff
@@ -0,0 +1,81 @@
+import copy
+from typing import List, Tuple, Union
+
+import numpy as np
+
+from torch_geometric.data import Data, HeteroData
+from torch_geometric.data.datapipes import functional_transform
+from torch_geometric.transforms import BaseTransform
+from torch_geometric.typing import NodeOrEdgeType
+
+
+@functional_transform('chronological_split')
+class ChronologicalSplit(BaseTransform):
+    r"""Performs a chronological split into training, validation and test sets
+    of a :class:`~torch_geometric.data.Data` or a
+    :class:`~torch_geometric.data.HeteroData` object
+    (functional name: :obj:`chronological_split`).
+
+    Chronological split satisfies the following relationship:
+    max(train_time) < min(val_time) && max(val_time) < min(test_time)
+
+    Args:
+        val_ratio (float, optional): The proportion (in percents) of the
+            dataset to include in the validation split.
+            (default: :obj:`0.15`)
+        test_ratio (float, optional): The proportion (in percents) of the
+            dataset to include in the test split. (default: :obj:`0.15`)
+    """
+    def __init__(self, val_ratio: float = 0.15,
+                 test_ratio: float = 0.15) -> None:
+        assert 0. <= val_ratio <= 1.
+        assert 0. <= test_ratio <= 1.
+        assert val_ratio + test_ratio < 1.
+        self.val_ratio = val_ratio
+        self.test_ratio = test_ratio
+        self.key = 'time'
+
+    def forward(
+        self,
+        data: Union[Data, HeteroData],
+    ) -> Tuple[
+            Union[Data, HeteroData],
+            Union[Data, HeteroData],
+            Union[Data, HeteroData],
+    ]:
+        if isinstance(data, Data):
+            return self._forward_data(data)
+        else:  # HeteroData
+            return self._forward_hetero_data(data)
+
+    def _forward_data(self, data: Data) -> Tuple[Data, Data, Data]:
+        assert hasattr(data, self.key)
+
+        val_time, test_time = np.quantile(
+            data[self.key].numpy(),
+            [1 - self.val_ratio - self.test_ratio, 1 - self.test_ratio])
+        train_data = data.snapshot(0, val_time)
+        val_data = data.snapshot(val_time, test_time)
```

Comment:
Snapshot return values in the period `[start, end]`. Given that both `train_data` and `val_data` can have data at `val_time`.

Reply:
This PR changes that behavior, take a look at [this line](https://github.com/pyg-team/pytorch_geometric/pull/8979/files#diff-46e4e6f655447088ed5da6e3517b0417723c19498cf31d0c28c781bea7e7d551R295). After re-thinking this I thought that right-excusive range is more natural.